### DISCUSSION TITLE
Contributing to MAIA

### DISCUSSION THREAD

**Discussion by botcs** [biigle/core#566] [Ideas]
Hi,

Thanks for maintaining and hosting BIIGLE, it's an awesome project!
I have been playing with the new panoptic segmentation model [SAM](https://ai.facebook.com/blog/segment-anything-foundation-model-image-segmentation/), and I was wondering if there was a way to contribute to MAIA, by swapping the Mask-RCNN to SAM.

Technically speaking my main question would be, how modular is the implementation of MAIA:
- super orthogonal: there's an interface that passes the patches from stage II in the [paper, fig1](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0207498) and expects nothing but bounding-boxes + class labels back?
- moderately hardwired: the model is initialized in the same process where the images are handled but is treated as a black box
- heavily baked together: only parts of the model is called with specific modifications to the MaskRCNN pipeline?

Possibly @mzur could give some insights from the #428 Hackathon? :)

## Why would MAIA benefit from SAM?
- Its zero shot capabilities could give highly valuable suggestions even without any prior labeling (see pic below)
- Finetuning the model on specific data requires a lot fewer examples than MaskRCNN
- The model is super-lightweight and the [github-repo of SAM](https://github.com/facebookresearch/segment-anything) already ships with ONNX porting
- With ONNX the model inference _can be_ made on the clients side, reducing maintenance cost on BIIGLE's side and improve accessibility to users.

![zero-shot segmentation](https://user-images.githubusercontent.com/12262395/230260015-1b61c553-afc0-4f7a-bb46-2b813acbb225.jpeg)

The pipeline itself from SAM is also intersting because it has a signficant overlap with MAIA, as both are aiming to refine the model using the annotators' output in an iterative way. This, however, would require a bit more of a research how could it be integrated in MAIA.

Thanks,
Csabi

**Comment by mzur** [comment]
Highly interesting! (@dlangenk you should take a look at this)

We recently replaced the TensorFlow Mask R-CNN implementation of MAIA with a PyTorch Faster R-CNN implementation (from [MMDetection](https://github.com/open-mmlab/mmdetection)). So the basic dependencies for SAM are already met. While we planned to replace Faster R-CNN with something better that is provided by MMDetection (which would be very easy), I'd be open to ditch MMDetection altogether if SAM is good enough. [This PR](https://github.com/biigle/maia/pull/117) can give you an idea which files must be touched if Faster R-CNN is to be replaced. It includes many insignificant changes in PHP files. The more relevant changes were done to the Python files. MAIA basically has one Python script to run the novelty detection, one to generate a training dataset (i.e. scaling images as part of the [UnKnoT method](https://doi.org/10.1109/ACCESS.2020.3014441)), one to run training on Faster R-CNN and one to run inference. Each script "communicates" with the PHP backend via JSON input/output files. It shouldn't be too hard to replace the scripts.

This needs careful evaluation, first, though. I see three possible applications for SAM in BIIGLE:

1. Be an alternative to the novelty detection of MAIA (or replace it if it is good enough). The zero-shot suggestions could be much better in more complicated images than the ones the novely detection expects. We should test SAM and compare the detection performance with the novelty detection on a few datasets. I wouldn't be surprised if SAM beats the novelty detection or at least has a much higher precision which already would be a big win.

2. Be a replacement for Faster R-CNN in MAIA. This must be tested on a few datasets as well. As a first step, SAM could be used as a plain (single-class) object detector with bounding boxes. This is what the whole MAIA workflow expects. Classification or segmentation would be great, too, but it would make the user interface and user interactions much more complicated (which will slow down the whole process again). We have a prototype UI that also includes training and manual filtering with class labels, though. We could test this with SAM. Ideas for segmentation support could come after that.

3. :sparkles: Most promising and low-hanging fruit: While I don't see an advantage for client-side inference with the two points above, client-side SAM could be used in the image annotation tool to replace or augment the magic wand (i.e. "hover and click"). It could also be used client-side for "segment everything" on a single image. This would need a new kind of workflow, though because BIIGLE has no concept for annotations without a label. But I'm sure we can come up with something.

So many ideas! Currently, we don't have the resources to work on this ourselves. It would be awesome if you would consider to contribute!

**Comment by dlangenk** [comment]
I had a look at the demo and tried it with a few images and it looks pretty impressive. Also I guess Nr 3 is the most promising one at the moment, but then it would be nice to have the same or similar user interface support as in the demo. Regarding the annotation without label this could easily be handled creating another global label called "object" or something similar. Will play with this a little bit more after easter. With the demo I guess we can quickly evaluate the potential for cases 1 and 2 as well.

**Comment by botcs** [comment]
Brilliant! thank you guys for the quick reply, I will look into the PR #117 to learn how to connect it to the interface.
Replying to @mzur 's 3 points:
1. would be an absolute uprooting, it would be nice to take the two models side-by-side: probs on MS-COCO we have some scores for fully supervised comparison, whereas LVIS shows few-shot scores of SAM. (Faster-RCNN needs some finetuning, so no clear comparison here) Check the main paper's Fig 9, Tab 3 and Tab 4 on these results
2. Originally I was thinking about this, but I haven't really thought about the UI consequences of switching from bounding boxes to segmentation
3. Yep. @dlangenk is right in that this seems to be a low-hanging fruit, that could massively speed up the annotation process right away.

**Comment by botcs** [comment]
Hi,

after looking into the PR, if I understood everything correctly I think the quickest update would be 2) modifying the following 2 files:
- [InferenceRunner.py](https://github.com/biigle/maia/blob/master/src/resources/scripts/object-detection/InferenceRunner.py)
- [TrainingRunner.py](https://github.com/biigle/maia/blob/master/src/resources/scripts/object-detection/TrainingRunner.py)

SAM does not come with a training script, so unless the pipeline actually makes use of the `class_index` (only occurence is L45 in the InferenceRunner.py), I don't see much point in finetuning the model on the downstream dataset, otherwise a tiny MLP layer could be implemented and optimized on the top of the segmented features?

----

On the other hand 3) would be really interesting to look into, is there a suggested way to set up a test environment for a new feature?

**Reply by dlangenk** [reply]
Regarding the test environment @mzur can provide better information.
I also don't think that we need to train on the actual data.
The main problem I see with 3) is that we need to compute the embedding for each image on the gpu (gpu takes about 3 sec per image (V100, Full HD image), but more than 30sec on the cpu), so we need to do this on the fly with some kind of queueing mechanism or precompute these and store them when the data is uploaded. The actual segmentation can then be done on the fly in the browser clientside with onnxruntime. The other problem is that we need some UI interactions which are missing at the moment, like selecting a point, showing the segmentation mask, selecting an additional positive or negative point, then show the segmentation mask again and finally submitting the annotation. Unfortunately the user interface of segmentAnything is not publicly available. Nevertheless this would be a massive improvment to Biigle in general.
I will discuss this with the group in the following days if the others have time for it.

**Reply by mzur** [reply]
@botcs Please have a look at the [developing instructions](https://github.com/biigle/core/blob/master/DEVELOPING.md) for how to set up a development instance. This new feature would probably be included directly in biigle/core (not in a module).

@dlangenk Computing an embedding on the fly on a GPU is no problem per se. We only have to allocate a whole GPU just for this. If we use the same GPU than for the MAIA jobs then it would be blocked for hours whenever a new MAIA job comes in. Your suggested alternative for precomputing the embeddings is also not bad. This is a longer running job (similar to MAIA) and could be queued on the same GPU. I would only compute the embeddings on request, though, i.e. the users have to press a button "prepare the volume for magic SAM" and then wait until they get notified that everything is ready.

As to the UI, we could start with the "single click segmentation" (i.e. what the demo does on mouse hover). Additional interactions (with positive/negative points) could come later. Everything is possible as long as the basic workflow of selecting a label first and then drawing the annotation stays intact. If the segmentation is not perfect, it can still be modified with the brush and eraser tools.

**Reply by dlangenk** [reply]
@mzur I have a kind of "working" prototype already. I'll show you when you return from illness/holiday.

**Comment by dlangenk** [comment]
I did some experiments with the segment everything part and I get not so good results at the moment. I created a script to download the images of a volume, apply the segmentAnything segment everything method and then upload the results to Biigle using Object as label. I attach the script here, but plan to make it available in the community-script repo later on. As I cannot attach py files here it is zipped additionally. I also tried different parameter sets, which changes the outcome quite a bit.
[segmentAnythingVolume.py.zip](https://github.com/biigle/core/files/11188964/segmentAnythingVolume.py.zip)

**Comment by mzur** [comment]
fyi: We are working on idea 3 and will release it soon. progress can be tracked in https://github.com/biigle/core/pull/584.
